In [26]:
import pandas as pd

In [27]:
pair_wise_out = pd.read_csv("../datasets/clustering FULL.csv") 
pair_wise_in = pd.read_csv("../datasets/Cluster Hit.csv")

In [28]:
pair_wise_out = pair_wise_out[pair_wise_out["WorkerId"] != "A2236R33CMX70Q"]
pair_wise_out = pair_wise_out[pair_wise_out["WorkerId"] != "A22B5CLRFQ9EUX"]

# Question:
# Don't Dream It's Over
# Sono come tu mi vuoi
# Don't Dream It's Over - The Best of Sixpence None the Richer
# Sixpence-None-the-Richer---Don't-Dream-It's-Over

# 1  3  4

In [29]:
true_labels = pair_wise_in.Label.values

In [30]:
(true_labels)

array(['[(1,3),(2,4)]', '[(1,3,4)]', '[(1,2)]', '[(1,4)]', '[(1,2,4)]',
       '[(1,2),(3,4)]', '[(1,3),(2,4)]', '[(1,3,4)]', '[(1,2,3)]',
       '[(1,3)]', '[(2,3,4)]', '[(1,2,3,4)]', '[(1,3),(2,4)]', '[(2,3)]',
       '[(3,4)]', '[(1,4),(2,3)]', '[(1,3),(2,4)]', '[(3,4)]',
       '[(1,2,3)]', '[(1,2),(3,4)]', '[(2,3)]', '[(1,2,3)]', '[]',
       '[(3,4)]', '[(1,3),(2,4)]', '[(1,2,3)]', '[(1,3),(2,4)]',
       '[(3,4)]', '[]', '[(1,2,3)]'], dtype=object)

In [31]:
#Needed to remove the ''
true_labels =[[(1,3),(2,4)], [(1,3,4)], [(1,2)], [(1,4)], [(1,2,4)],
       [(1,2),(3,4)], [(1,3),(2,4)], [(1,3,4)], [(1,2,3)],
       [(1,3)], [(2,3,4)], [(1,2,3,4)], [(1,3),(2,4)], [(2,3)],
       [(3,4)], [(1,4),(2,3)], [(1,3),(2,4)], [(3,4)],
       [(1,2,3)], [(1,2),(3,4)], [(2,3)], [(1,2,3)], [],
       [(3,4)], [(1,3),(2,4)], [(1,2,3)], [(1,3),(2,4)],
       [(3,4)], [], [(1,2,3)]]

In [32]:
pair_wise_out['Answer.taskAnswers'].values

array(['[{"label_1":"1","label_2":"2","label_3":"1","label_4":"2"}]',
       '[{"label_1":"1","label_2":"2","label_3":"1","label_4":"2"}]',
       '[{"label_1":"1","label_2":"2","label_3":"1","label_4":"2"}]',
       '[{"label_1":"1","label_2":"2","label_3":"1","label_4":"2"}]',
       '[{"label_1":"1","label_2":"2","label_3":"1","label_4":"2"}]',
       '[{"label_1":"1","label_2":"2","label_3":"1","label_4":"1"}]',
       '[{"label_1":"1","label_2":"2","label_3":"1","label_4":"1"}]',
       '[{"label_1":"1","label_2":"2","label_3":"1","label_4":"1"}]',
       '[{"label_1":"3","label_2":"4","label_3":"2","label_4":"1"}]',
       '[{"label_1":"1","label_2":"2","label_3":"1","label_4":"1"}]',
       '[{"label_1":"1","label_2":"1","label_3":"2","label_4":"3"}]',
       '[{"label_1":"1","label_2":"4","label_3":"2","label_4":"3"}]',
       '[{"label_1":"1","label_2":"1","label_3":"2","label_4":"3"}]',
       '[{"label_1":"1","label_2":"1","label_3":"3","label_4":"2"}]',
       '[{"label_1":

In [33]:
outty=[[{"1":"1","2":"2","3":"1","4":"2"}],
       [{"1":"1","2":"2","3":"1","4":"2"}],
       [{"1":"1","2":"2","3":"1","4":"2"}],
       [{"1":"1","2":"2","3":"1","4":"2"}],
       [{"1":"1","2":"2","3":"1","4":"2"}],
       [{"1":"1","2":"2","3":"1","4":"1"}],
       [{"1":"1","2":"2","3":"1","4":"1"}],
       [{"1":"1","2":"2","3":"1","4":"1"}],
       [{"1":"3","2":"4","3":"2","4":"1"}],
       [{"1":"1","2":"2","3":"1","4":"1"}],
       [{"1":"1","2":"1","3":"2","4":"3"}],
       [{"1":"1","2":"4","3":"2","4":"3"}],
       [{"1":"1","2":"1","3":"2","4":"3"}],
       [{"1":"1","2":"1","3":"3","4":"2"}],
       [{"1":"1","2":"1","3":"2","4":"3"}],
       [{"1":"1","2":"2","3":"3","4":"1"}],
       [{"1":"1","2":"2","3":"3","4":"1"}],
       [{"1":"1","2":"2","3":"3","4":"1"}],
       [{"1":"1","2":"2","3":"3","4":"1"}],
       [{"1":"1","2":"2","3":"3","4":"1"}],
       [{"1":"1","2":"1","3":"2","4":"1"}],
       [{"1":"1","2":"1","3":"2","4":"1"}],
       [{"1":"1","2":"1","3":"2","4":"1"}],
       [{"1":"1","2":"1","3":"2","4":"1"}],
       [{"1":"1","2":"1","3":"2","4":"1"}],
       [{"1":"2","2":"2","3":"1","4":"1"}],
       [{"1":"1","2":"1","3":"2","4":"2"}],
       [{"1":"1","2":"1","3":"2","4":"2"}],
       [{"1":"1","2":"1","3":"2","4":"2"}],
       [{"1":"1","2":"1","3":"2","4":"2"}],
       [{"1":"1","2":"2","3":"1","4":"2"}],
       [{"1":"1","2":"2","3":"1","4":"2"}],
       [{"1":"1","2":"2","3":"1","4":"2"}],
       [{"1":"1","2":"2","3":"1","4":"2"}],
       [{"1":"1","2":"2","3":"1","4":"2"}],
       [{"1":"1","2":"2","3":"1","4":"1"}],
       [{"1":"1","2":"2","3":"1","4":"1"}],
       [{"1":"1","2":"2","3":"1","4":"1"}],
       [{"1":"1","2":"2","3":"1","4":"1"}],
       [{"1":"1","2":"2","3":"1","4":"1"}],
       [{"1":"1","2":"1","3":"1","4":"2"}],
       [{"1":"1","2":"1","3":"1","4":"2"}],
       [{"1":"1","2":"1","3":"1","4":"2"}],
       [{"1":"1","2":"1","3":"1","4":"2"}],
       [{"1":"1","2":"2","3":"3","4":"4"}],
       [{"1":"1","2":"2","3":"3","4":"4"}],
       [{"1":"1","2":"2","3":"1","4":"3"}],
       [{"1":"1","2":"3","3":"2","4":"4"}],
       [{"1":"1","2":"2","3":"1","4":"4"}],
       [{"1":"1","2":"2","3":"1","4":"3"}],
       [{"1":"1","2":"2","3":"2","4":"2"}],
       [{"1":"1","2":"2","3":"2","4":"2"}],
       [{"1":"2","2":"1","3":"1","4":"1"}],
       [{"1":"1","2":"2","3":"2","4":"2"}],
       [{"1":"1","2":"2","3":"2","4":"2"}],
       [{"1":"1","2":"1","3":"1","4":"1"}],
       [{"1":"1","2":"1","3":"1","4":"1"}],
       [{"1":"1","2":"1","3":"1","4":"1"}],
       [{"1":"1","2":"1","3":"1","4":"1"}],
       [{"1":"1","2":"1","3":"1","4":"1"}],
       [{"1":"1","2":"2","3":"1","4":"2"}],
       [{"1":"1","2":"2","3":"1","4":"2"}],
       [{"1":"1","2":"2","3":"1","4":"2"}],
       [{"1":"1","2":"2","3":"1","4":"2"}],
       [{"1":"1","2":"2","3":"1","4":"2"}],
       [{"1":"1","2":"2","3":"2","4":"3"}],
       [{"1":"1","2":"2","3":"2","4":"3"}],
       [{"1":"3","2":"1","3":"1","4":"2"}],
       [{"1":"2","2":"1","3":"1","4":"2"}],
       [{"1":"1","2":"2","3":"2","4":"3"}],
       [{"1":"3","2":"2","3":"1","4":"1"}],
       [{"1":"1","2":"2","3":"3","4":"3"}],
       [{"1":"3","2":"1","3":"2","4":"2"}],
       [{"1":"1","2":"2","3":"3","4":"3"}],
       [{"1":"1","2":"2","3":"3","4":"3"}],
       [{"1":"1","2":"2","3":"2","4":"1"}],
       [{"1":"2","2":"1","3":"1","4":"2"}],
       [{"1":"1","2":"2","3":"2","4":"1"}],
       [{"1":"1","2":"2","3":"2","4":"1"}],
       [{"1":"1","2":"3","3":"4","4":"2"}],
       [{"1":"1","2":"2","3":"1","4":"2"}],
       [{"1":"2","2":"1","3":"2","4":"1"}],
       [{"1":"1","2":"2","3":"1","4":"2"}],
       [{"1":"1","2":"2","3":"1","4":"2"}],
       [{"1":"1","2":"2","3":"1","4":"2"}],
       [{"1":"2","2":"3","3":"1","4":"1"}],
       [{"1":"3","2":"2","3":"1","4":"1"}],
       [{"1":"2","2":"2","3":"1","4":"1"}],
       [{"1":"1","2":"2","3":"3","4":"3"}],
       [{"1":"3","2":"2","3":"1","4":"1"}],
       [{"1":"1","2":"1","3":"1","4":"2"}],
       [{"1":"1","2":"1","3":"1","4":"2"}],
       [{"1":"1","2":"1","3":"1","4":"2"}],
       [{"1":"1","2":"1","3":"1","4":"2"}],
       [{"1":"1","2":"1","3":"1","4":"2"}],
       [{"1":"2","2":"2","3":"1","4":"1"}],
       [{"1":"1","2":"1","3":"2","4":"2"}],
       [{"1":"1","2":"1","3":"2","4":"2"}],
       [{"1":"4","2":"1","3":"2","4":"3"}],
       [{"1":"1","2":"1","3":"2","4":"2"}],
       [{"1":"1","2":"2","3":"3","4":"4"}],
       [{"1":"3","2":"1","3":"1","4":"2"}],
       [{"1":"1","2":"2","3":"2","4":"3"}],
       [{"1":"1","2":"2","3":"3","4":"4"}],
       [{"1":"3","2":"1","3":"1","4":"2"}],
       [{"1":"1","2":"1","3":"1","4":"2"}],
       [{"1":"1","2":"1","3":"1","4":"2"}],
       [{"1":"2","2":"1","3":"1","4":"3"}],
       [{"1":"1","2":"1","3":"1","4":"4"}],
       [{"1":"1","2":"1","3":"1","4":"2"}],
       [{"1":"1","2":"2","3":"3","4":"4"}],
       [{"1":"1","2":"2","3":"3","4":"4"}],
       [{"1":"1","2":"2","3":"3","4":"4"}],
       [{"1":"1","2":"2","3":"3","4":"4"}],
       [{"1":"1","2":"2","3":"4","4":"3"}],
       [{"1":"3","2":"2","3":"1","4":"1"}],
       [{"1":"1","2":"2","3":"3","4":"3"}],
       [{"1":"1","2":"2","3":"3","4":"3"}],
       [{"1":"1","2":"2","3":"3","4":"3"}],
       [{"1":"3","2":"2","3":"1","4":"1"}],
       [{"1":"1","2":"2","3":"1","4":"2"}],
       [{"1":"1","2":"2","3":"1","4":"2"}],
       [{"1":"1","2":"2","3":"1","4":"2"}],
       [{"1":"1","2":"2","3":"1","4":"2"}],
       [{"1":"1","2":"2","3":"1","4":"2"}],
       [{"1":"1","2":"2","3":"2","4":"3"}],
       [{"1":"1","2":"1","3":"1","4":"2"}],
       [{"1":"1","2":"1","3":"1","4":"2"}],
       [{"1":"1","2":"1","3":"1","4":"2"}],
       [{"1":"1","2":"1","3":"1","4":"2"}],
       [{"1":"1","2":"2","3":"1","4":"2"}],
       [{"1":"2","2":"1","3":"2","4":"1"}],
       [{"1":"1","2":"2","3":"1","4":"2"}],
       [{"1":"1","2":"2","3":"1","4":"2"}],
       [{"1":"1","2":"2","3":"1","4":"2"}],
       [{"1":"3","2":"4","3":"2","4":"1"}],
       [{"1":"3","2":"2","3":"1","4":"1"}],
       [{"1":"3","2":"2","3":"1","4":"1"}],
       [{"1":"1","2":"2","3":"3","4":"3"}],
       [{"1":"1","2":"2","3":"3","4":"3"}],
       [{"1":"3","2":"1","3":"2","4":"4"}],
       [{"1":"1","2":"2","3":"3","4":"4"}],
       [{"1":"1","2":"2","3":"3","4":"4"}],
       [{"1":"1","2":"2","3":"3","4":"4"}],
       [{"1":"1","2":"2","3":"3","4":"4"}],
       [{"1":"1","2":"1","3":"1","4":"2"}],
       [{"1":"1","2":"1","3":"1","4":"2"}],
       [{"1":"1","2":"1","3":"1","4":"2"}],
       [{"1":"1","2":"1","3":"1","4":"2"}],
       [{"1":"1","2":"1","3":"2","4":"3"}]]

In [34]:
len(outty)

150

In [35]:
worker_labels=[]
for i in range(len(outty)):
    clusters=[]
    for j in list(pd.Series(list(outty[i][0].values())).value_counts().index):
        mini_clusters=[]
        for k in list(outty[i][0].keys()):
            if outty[i][0][k]==j:
                mini_clusters.append(int(k))
        if len(mini_clusters)>1:        
            clusters.append(tuple(mini_clusters))
    worker_labels.append(clusters)

In [36]:
len(worker_labels)

150

In [42]:
import numpy as np
true_labelss = [[e] * 5 for e in true_labels]
true_labelss=np.array(true_labelss).ravel()


In [43]:
worker_labels

[[(1, 3), (2, 4)],
 [(1, 3), (2, 4)],
 [(1, 3), (2, 4)],
 [(1, 3), (2, 4)],
 [(1, 3), (2, 4)],
 [(1, 3, 4)],
 [(1, 3, 4)],
 [(1, 3, 4)],
 [],
 [(1, 3, 4)],
 [(1, 2)],
 [],
 [(1, 2)],
 [(1, 2)],
 [(1, 2)],
 [(1, 4)],
 [(1, 4)],
 [(1, 4)],
 [(1, 4)],
 [(1, 4)],
 [(1, 2, 4)],
 [(1, 2, 4)],
 [(1, 2, 4)],
 [(1, 2, 4)],
 [(1, 2, 4)],
 [(3, 4), (1, 2)],
 [(1, 2), (3, 4)],
 [(1, 2), (3, 4)],
 [(1, 2), (3, 4)],
 [(1, 2), (3, 4)],
 [(1, 3), (2, 4)],
 [(1, 3), (2, 4)],
 [(1, 3), (2, 4)],
 [(1, 3), (2, 4)],
 [(1, 3), (2, 4)],
 [(1, 3, 4)],
 [(1, 3, 4)],
 [(1, 3, 4)],
 [(1, 3, 4)],
 [(1, 3, 4)],
 [(1, 2, 3)],
 [(1, 2, 3)],
 [(1, 2, 3)],
 [(1, 2, 3)],
 [],
 [],
 [(1, 3)],
 [],
 [(1, 3)],
 [(1, 3)],
 [(2, 3, 4)],
 [(2, 3, 4)],
 [(2, 3, 4)],
 [(2, 3, 4)],
 [(2, 3, 4)],
 [(1, 2, 3, 4)],
 [(1, 2, 3, 4)],
 [(1, 2, 3, 4)],
 [(1, 2, 3, 4)],
 [(1, 2, 3, 4)],
 [(1, 3), (2, 4)],
 [(1, 3), (2, 4)],
 [(1, 3), (2, 4)],
 [(1, 3), (2, 4)],
 [(1, 3), (2, 4)],
 [(2, 3)],
 [(2, 3)],
 [(2, 3)],
 [(2, 3), (1, 4)],
 [(2

In [44]:
TP=[]
FP=[]
FN=[]
TN=[]


for i,j in zip(worker_labels,true_labelss):
    
    tp=len(set(i).intersection(j))
    fp=len(set(i)-set(j))
    fn=len(set(j)-set(i))
    tn=256-(tp+fp+fn)
    TP.append(tp)
    FP.append(fp)
    FN.append(fn)
    TN.append(tn)
    
    

In [45]:
(np.array(TP).mean()),(np.array(FP).mean()),(np.array(TN).mean()),(np.array(FN).mean())

(1.1333333333333333, 0.03333333333333333, 254.73333333333332, 0.1)

In [46]:
prec= (np.array(TP).mean())/((np.array(TP).mean())+(np.array(FP).mean()))
prec

0.9714285714285713

In [47]:
recall=(np.array(TP).mean())/((np.array(TP).mean())+(np.array(FN).mean()))
recall

0.9189189189189189

In [48]:
F1_score = 2* (prec*recall)/(prec+recall)
F1_score

0.9444444444444444

# With honeypot:
#### precision = 0.97142857142857
#### recall = 0.91891891
#### f1 = 0.9444444

# Without honeypot:
#### precision = 0.9547
#### recall = 0.8757
#### f1 = 0.9243

In [18]:
single_label_7=[]
for i in range(0,len(worker_labels),7):
    single_label_7.append(pd.Series(worker_labels[i:i+7]).value_counts().index[0])

In [19]:
single_label_7

[[(1, 3), (2, 4)],
 [(1, 3, 4)],
 [(1, 2)],
 [(1, 4)],
 [(1, 2, 4)],
 [(1, 2), (3, 4)],
 [(1, 3), (2, 4)],
 [(1, 3, 4)],
 [(1, 2, 3)],
 [(1, 3)],
 [(2, 3, 4)],
 [(1, 2, 3, 4)],
 [(1, 3), (2, 4)],
 [(2, 3)],
 [(3, 4)],
 [(1, 4), (2, 3)],
 [(1, 3), (2, 4)],
 [(3, 4)],
 [(1, 2, 3)],
 [(1, 2), (3, 4)],
 [(2, 3)],
 [(1, 2, 3)],
 [],
 [(3, 4)],
 [(1, 3), (2, 4)],
 [(1, 2, 3)],
 [(1, 3), (2, 4)],
 [(3, 4)],
 [],
 [(1, 2, 3)]]

In [20]:
#As you can see they're all correct. F1, Precision, Recall is 1. 
true_labels

[[(1, 3), (2, 4)],
 [(1, 3, 4)],
 [(1, 2)],
 [(1, 4)],
 [(1, 2, 4)],
 [(1, 2), (3, 4)],
 [(1, 3), (2, 4)],
 [(1, 3, 4)],
 [(1, 2, 3)],
 [(1, 3)],
 [(2, 3, 4)],
 [(1, 2, 3, 4)],
 [(1, 3), (2, 4)],
 [(2, 3)],
 [(3, 4)],
 [(1, 4), (2, 3)],
 [(1, 3), (2, 4)],
 [(3, 4)],
 [(1, 2, 3)],
 [(1, 2), (3, 4)],
 [(2, 3)],
 [(1, 2, 3)],
 [],
 [(3, 4)],
 [(1, 3), (2, 4)],
 [(1, 2, 3)],
 [(1, 3), (2, 4)],
 [(3, 4)],
 [],
 [(1, 2, 3)]]

In [22]:
errors=0
for i,j in zip(true_labels,single_label_7):
    errors+=len(set(i)-set(i))
#     input(i)
#     input(j)
#     input(errors)

In [23]:
errors

0

In [ ]:
#As you can see they're all correct. F1, Precision, Recall is 1. 